File: src/utils/security.ts
import * as bcrypt from 'bcryptjs';
import { v4 as uuidv4 } from 'uuid';
import jwt from '@tsndr/cloudflare-worker-jwt';
import type { JwtPayload, QualityLevel } from '../types';

const SALT_ROUNDS = 10;

export async function hashPassword(password: string): Promise<string> {
  const salt = await bcrypt.genSalt(SALT_ROUNDS);
  return bcrypt.hash(password, salt);
}

export async function verifyPassword(password: string, hash: string): Promise<boolean> {
  return bcrypt.compare(password, hash);
}

export function generateUUID(): string {
  return uuidv4();
}

export function generateSessionId(): string {
  const bytes = new Uint8Array(32);
  crypto.getRandomValues(bytes);
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

export function generateFileId(): string {
  const bytes = new Uint8Array(9);
  crypto.getRandomValues(bytes);
  return btoa(String.fromCharCode(...bytes))
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

export function generateSecretToken(): string {
  const bytes = new Uint8Array(48);
  crypto.getRandomValues(bytes);
  return btoa(String.fromCharCode(...bytes))
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

export async function createAccessToken(
  username: string,
  secretKey: string,
  expiresInMinutes: number
): Promise<string> {
  const now = Math.floor(Date.now() / 1000);
  const payload: JwtPayload = {
    sub: username,
    iat: now,
    exp: now + expiresInMinutes * 60,
  };
  
  return jwt.sign(payload, secretKey, {
    algorithm: 'HS256',
  });
}

export async function verifyAccessToken(
  token: string,
  secretKey: string
): Promise<JwtPayload | null> {
  try {
    const isValid = await jwt.verify(token, secretKey, {
      algorithm: 'HS256',
    });
    
    if (!isValid) return null;
    
    const decoded = jwt.decode(token) as { payload: JwtPayload };
    return decoded.payload;
  } catch {
    return null;
  }
}

export function hashFilename(filename: string): string {
  const timestamp = Date.now().toString();
  const randomBytes = new Uint8Array(16);
  crypto.getRandomValues(randomBytes);
  const randomStr = Array.from(randomBytes, (b) => b.toString(16).padStart(2, '0')).join('');
  
  const hashInput = `${filename}${timestamp}${randomStr}`;
  const encoder = new TextEncoder();
  const data = encoder.encode(hashInput);
  
  return Array.from(new Uint8Array(data.slice(0, 32)), (b) => b.toString(16).padStart(2, '0')).join('');
}

export const QUALITY_CONFIGS: Record<QualityLevel, { maxWidth: number; maxHeight: number; quality: number; suffix: string }> = {
  original: { maxWidth: 0, maxHeight: 0, quality: 100, suffix: 'original' },
  standard: { maxWidth: 1920, maxHeight: 1080, quality: 85, suffix: 'standard' },
  low: { maxWidth: 800, maxHeight: 600, quality: 60, suffix: 'low' },
  icon: { maxWidth: 200, maxHeight: 200, quality: 70, suffix: 'icon' },
};

export function parseAllowedTypes(typesStr: string): string[] {
  return typesStr.split(',').map((t) => t.trim());
}

export function getFileExtension(filename: string): string {
  const ext = filename.split('.').pop();
  return ext ? ext.toLowerCase() : '';
}

File: app/utils/security.py
import base64
import hashlib
import secrets
from datetime import datetime, timedelta
from typing import Optional

import bcrypt
from jose import JWTError, jwt

from app.config import get_settings

settings = get_settings()


def _preprocess_password(password: str) -> bytes:
    password_bytes = password.encode("utf-8")
    hashed = hashlib.sha256(password_bytes).digest()
    return base64.b64encode(hashed)


def verify_password(plain_password: str, hashed_password: str) -> bool:
    password_bytes = _preprocess_password(plain_password)
    hashed_bytes = hashed_password.encode("utf-8")
    return bcrypt.checkpw(password_bytes, hashed_bytes)


def get_password_hash(password: str) -> str:
    password_bytes = _preprocess_password(password)
    salt = bcrypt.gensalt()
    hashed = bcrypt.hashpw(password_bytes, salt)
    return hashed.decode("utf-8")


def create_access_token(data: dict, expires_delta: Optional[timedelta] = None) -> str:
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode.update({"exp": expire})
    encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
    return encoded_jwt


def decode_access_token(token: str) -> Optional[dict]:
    try:
        payload = jwt.decode(token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM])
        return payload
    except JWTError:
        return None


def generate_session_id() -> str:
    return secrets.token_hex(32)


def generate_file_id() -> str:
    return secrets.token_urlsafe(12)


def hash_filename(filename: str) -> str:
    timestamp = datetime.utcnow().strftime("%Y%m%d%H%M%S%f")
    random_str = secrets.token_hex(8)
    hash_input = f"{filename}{timestamp}{random_str}".encode()
    return hashlib.sha256(hash_input).hexdigest()[:32]
